### PR TITLE
OOP school library: unit tests

### PR DESCRIPTION
**Pull request by nilab-khwaja**
Add unit tests for all the methods in all the classes:

- Book
- Person
- Student
- Teacher
- Classroom
- Rental
- all classes that you created while implementing the Decorator design pattern

### CODE REVIEW

**Review by Gambit142 — COMMENTED** [review]
Hi @nilab-khwaja @Izebiafe 👋, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

- The rubocop linters checks on GitHub failed. Kindly correct the linters errors. 👍 😎 

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: specs/book_spec.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+      allow(person).to receive(:add_rental)
+      book = Book.new('Title', 'Author')
+      date = '2023-01-01'
+
+      book.add_rental(person, date)
+
+      expect(book.rentals).not_to be_empty
+      expect(person).to have_received(:add_rental).with(an_instance_of(Rental))
+    end
+  end
```
Comment (danmainah) [inline specs/book_spec.rb:37]: - [ ] @nilab-khwaja and @Izebiafe Kindly consider adding tests for all methods in the Book class, you are missing tests for `initialize` and `to_json`, kindly consider adding them:

image

File: specs/person_spec.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      expect(result).to be true
+    end
+
+    it 'returns true if the person has parent permission' do
+      person = Person.new(16, 'John', parent_permission: true)
+
+      result = person.can_use_services?
+
+      expect(result).to be true
+    end
```
Comment (danmainah) [inline specs/person_spec.rb:19]: - [ ] @nilab-khwaja and @Izebiafe Kindly consider adding tests for all methods in the Person class, you are missing tests for `correct_name`, `add_rental` and `to_json`, kindly consider adding them:

image

File: specs/decorator_spec.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  let(:nameable_mock) { double('Nameable') }
+  subject { Decorator.new(nameable_mock) }
+
+  describe '#correct_name' do
+    it 'delegates correct_name to the wrapped Nameable object' do
+      expect(nameable_mock).to receive(:correct_name)
+      subject.correct_name
+    end
+  end
+end
```
Comment (danmainah) [inline specs/decorator_spec.rb:13]: - [ ] @Team Great work adding tests for the decoration classes however you forgot to add tests for the  Nameable class, kindly consider adding test for it.

image

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @nilab-khwaja and @Izebiafe , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️  Added tests for all required classes except one.
✔️  Ensured the tests added are all passing


# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by danmainah — APPROVED** [review]
Hi @nilab-khwaja and @Izebiafe , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 

## Highlights
✔️ Implemented the suggested changes to ensure all tests are passing

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**